lib/plugins/aws/package/compile/events/sns/index.js

```diff
@@ -11,38 +11,49 @@ class AwsCompileSNSEvents {
       'package:compileEvents': this.compileSNSEvents.bind(this),
     };
 
-    // TODO: Complete schema, see https://github.com/serverless/serverless/issues/8035
     this.serverless.configSchemaHandler.defineFunctionEvent('aws', 'sns', {
-      anyOf: [{ type: 'string' }, { type: 'object' }],
+      oneOf: [
+        { type: 'string', maxLength: 256, pattern: '^[\\w-]+$' },
+        { $ref: '#/definitions/awsArn' },
+        {
+          type: 'object',
+          properties: {
+            arn: {
+              oneOf: [{ $ref: '#/definitions/awsArn' }, { $ref: '#/definitions/awsCfFunction' }],
+            },
```

Comment:
What do you think about renaming `awsArn` definition (in current form) to `awsArnString`, and introduce `awsArn` definition, that allows form as accepted by AWS, which then we can use here directly?

Reply:
What I don't like too much is that, while the current `awsArn` ensure we're talking about an ARN, there are no guarantee that value returned by any of the `awsCfFunction` are actually resolved to ARN. I could use a GetAtt to get Lambda current version and this will pass the new `awsArn` definition.

More largely, any value within a CF template can be replaced by an AWS function resolution.

What we could envision here, is whenever one value used in `serverless.yaml` will be passed as is to a CF template, this value can either be the expected primitive type of the CF template (`string`, `number`, `boolean`) as well as any combination of CF functions.

We could then have :

```js
awsString: {
  oneOf: [{ type: 'string', minLength: 1}, { $ref: '#/definitions/awsCfFunction' }]
},
awsNumber: ...
```

WDYT ?

Reply:
> there are no guarantee that value returned by any of the awsCfFunction are actually resolved to ARN.

My idea for `awsArn` is that it's expected to be a valid ARN passed as is to AWS, which we validate just on surface, and it's AWS that eventually crashes with meaningful error, when nonsense instruction is passed.

I was thinking about keeping this as definition, as I believe we have many places where we propagate ARN down to a CF template supporting all instruction forms and without thoroughly validating it. Having that this definition could well serve all those place. (I think it's better to have reusable definition for that, instead being required to put `oneOf` combined instruction as above in all  of them)

> will be passed as is to a CF template, this value can either be the expected primitive type of the CF template

One problem of that approach is that we loose some sanity validation of _string_ version, which could be nice to have

So I wonder whether it's not better to have dedicated definitions for types that are used frequently (like ARN) and in other cases use `oneOf` directly as you've proposed above.

Reply:
understood @medikoo , updated accordingly :)